python/ray/air/execution/resources/resource_manager.py

```diff
@@ -0,0 +1,123 @@
+
+
+@DeveloperAPI
+class ResourceManager(abc.ABC):
+    """Resource manager interface.
+
+    A resource manager can be used to request resources from a Ray cluster and
+    allocate them to remote Ray tasks.
+
+    Because Ray clusters can autoscale, we need to request resources before
+    we can allocate them. Effectively this means we have to keep track
+    of requested resources (which can trigger autoscaling). Once the requested
+    resources are available to be scheduled, the control loop can acquire them.
+
+    This separation also ensures that requests can be cancelled before they resolve
+    if resources are not needed anymore.
+
+    Thus, resources managed by the resource manager have three states:
+
+    1. Requested but not available (ready). The resource request is received,
+       but the resources are not available for use of remote Ray objects, yet.
+    2. Requested and ready. The resources can be acquired and subsequently used by
+       remote Ray objects.
+    3. Acquired. These resources have been acquired by a caller.
+
+    The flow is as follows:
+
+    .. code-block:: python
+
+        # Create resource manager
+        resource_manager = ResourceManager()
+
+        # Create resource request
+        resource_request = ResourceRequest([{"CPU": 4}])
+
+        # Pass to resource manager
+        resource_manager.request_resources(resource_request)
+
+        # Wait until ready
+        while not resource_manager.has_resources_ready(resource_request):
+            time.sleep(1)
+
+        # Once ready, acquire resources
+        acquired_resource = resource_manager.acquire_resources(resource_request)
+
+        # Bind to remote task or actor
+        annotated_remote_fn = acquired_resource.annotate_remote_objects(
+            [remote_fn])
+
+        # Run remote function. This will use the acquired resources
+        ray.get(annotated_remote_fn.remote())
+
+        # After using the resources, free
+        resource_manager.free_resources(annotated_resources)
+
+    """
+
+    def request_resources(self, resource_request: ResourceRequest):
+        """Request resources, e.g. schedule placement group."""
+        raise NotImplementedError
+
+    def cancel_resource_request(self, resource_request: ResourceRequest):
+        """Cancel resource request, e.g. remove placement group."""
+        raise NotImplementedError
+
+    def has_resources_ready(self, resource_request: ResourceRequest) -> bool:
+        """Returns True if resources for the given request are available"""
+        raise NotImplementedError
+
+    def acquire_resources(
+        self, resource_request: ResourceRequest
+    ) -> Optional[AcquiredResource]:
+        """Acquire resources. Returns None if resources are not available."""
+        raise NotImplementedError
+
+    def free_resources(self, acquired_resource: AcquiredResource):
+        """Free resources from usage."""
+        raise NotImplementedError
```

Comment:
Can we clarify in the interface what the difference is between "cancel", "free", and "clear"? This will help make sure the expectations are matched across different implementations of this interface.

My understanding is:
- **Cancel:** Cancel a resource that is "requested" or "ready".
- **Free:** Free a resource that is "scheduled" and return it back to the cluster.
- **Clear:** Cancel all "requested" and "ready" resources, and free all "scheduled" resources.

Reply:
can we write some good docstrings please?